source/common/router/config_impl.cc

```diff
@@ -136,6 +137,37 @@ Optional<uint64_t> HashPolicyImpl::generateHash(const std::string& downstream_ad
   return hash;
 }
 
+std::vector<MetadataMatchCriterionConstSharedPtr>
+MetadataMatchCriteriaImpl::extractMetadataMatchCriteria(const MetadataMatchCriteriaImpl* parent,
```

Comment:
Can we document somewhere (maybe constructor) the significance of parent-child relationship of `MetadataMatchCriteria`? I think this bit is non-intuitive, the mechanics seem fine though.

Reply:
I think the commit message actually has this info, so probably just inline it into some comment as well.

Reply:
I did some of this, but let me refactor it a bit. I think it'll make more sense if the merging happens in an explicit call to a member function on the parent instead of using a special constructor.